src/webresource-handler/index.ts

```diff
@@ -216,6 +249,136 @@ export const getUploaderMiddlware = (
 };
 
+const parseUploadParameters = (
+	params: any,
+): WebResourceUploadParameters | null => {
+	const { vocabulary, resourceName, id, fieldName } = params;
+	if (
+		typeof vocabulary !== 'string' ||
+		typeof resourceName !== 'string' ||
+		typeof fieldName !== 'string' ||
+		isNaN(Number(id))
+	) {
+		return null;
+	}
+	const parsedId: number = Number(id);
+	return {
+		vocabulary,
+		resourceName,
+		id: parsedId,
+		fieldName,
+	};
+};
+
+const validateRequestBody = (body: any): WebResourceUploadPayload | null => {
+	const { metadata } = body;
+	if (
+		metadata == null ||
+		typeof metadata !== 'object' ||
+		typeof metadata.size !== 'number' ||
+		typeof metadata.filename !== 'string' ||
+		typeof metadata.content_type !== 'string'
+	) {
+		return null;
+	}
+	return {
+		metadata,
+	};
+};
+
+const canAccess = async (
+	uploadParameters: WebResourceUploadParameters,
+): Promise<boolean> => {
+	const { resourceName, vocabulary, id } = uploadParameters;
+	try {
+		await sbvrUtils.api[vocabulary].post({
+			url: `${resourceName}(${id})/canAccess`,
+			body: { action: 'update' },
+		});
+
+		return true;
+	} catch (err) {
+		getLogger(vocabulary).warn(
+			'Failed to validate access for',
+			resourceName,
+			id,
+		);
+		return false;
+	}
+};
+
+const getUploadUrl: Express.RequestHandler = async (req, res, next) => {
+	const handler = getWebresourceHandler();
+	if (handler == null) {
+		return next(
+			new errors.BadRequestError('No webresource handler configured'),
+		);
+	}
+
+	const uploadParameters = parseUploadParameters(req.query);
+	if (uploadParameters == null) {
+		return next(new errors.BadRequestError('Invalid request parameters'));
+	}
+
+	const requestBody = validateRequestBody(req.body);
+	if (requestBody == null) {
+		return next(new errors.BadRequestError('Invalid request body'));
+	}
+
+	const hasPermissions = await canAccess(uploadParameters);
+	if (!hasPermissions) {
+		return next(new errors.ForbiddenError());
+	}
+
+	const { metadata } = requestBody;
+	const { presignedUrl, token } = await handler.getUploadUrl(
+		uploadParameters,
+		metadata,
+	);
+	res.status(200).json({ presignedUrl, token });
+};
+
+const commitUpload: Express.RequestHandler = async (req, res, next) => {
+	const handler = getWebresourceHandler();
+	if (handler == null) {
+		return next(
+			new errors.BadRequestError('No webresource handler configured'),
+		);
+	}
+
+	const token = req.body.token;
+	if (token == null) {
+		return next(new errors.BadRequestError('Upload token is required'));
+	}
+
+	const decodedPayload = await handler.decodeUploadToken(token);
+	const hasPermissions = await canAccess(decodedPayload.uploadParameters);
+	if (!hasPermissions) {
+		return next(new errors.ForbiddenError());
+	}
+	// Note that this has to be done in this order, first commit on the storage and then on the DB
+	// to ensure that we will only store metadata if the file is actually stored
+	const webresource = await handler.commitUpload(decodedPayload);
```

Comment:
You may have to provisionally update the db first in order to prevent overlapping operations (how would the system behave with multiple concurrent calls to “commit”?), then move the file, then finally update the db to finalize the upload.

Reply:
I am not sure this is needed @dfunckt as it adds some (imo unnecessary) state to the DB.
In the case of multiple concurrent calls to commit, as long as one successfully copy the file and then delete it (and even if multiple copies one over the other, this is the same file being copied) after any succeeds to delete the staging file, the subsequent commits will fail. Or even if the user re-uploads to the presigned URL while commiting, at the end we would still not have failed data integrity as the metadata stored on DB would be pointing to the latest upl